pulsar-broker-common/src/main/java/org/apache/pulsar/common/configuration/VipStatus.java

```diff
@@ -55,13 +64,27 @@ public String checkStatus() {
             File statusFile = new File(statusFilePath);
             if (isReady && statusFile.exists() && statusFile.isFile()) {
                 // check deadlock
-                String diagnosticResult = ThreadDumpUtil.buildThreadDiagnosticString();
-                if (StringUtils.isBlank(diagnosticResult)) {
-                    return "OK";
-                } else {
-                    log.warn("Deadlock detected, service may be unavailable, "
-                            + "thread stack details are as follows: {}.", diagnosticResult);
+                ThreadMXBean threadBean = ManagementFactory.getThreadMXBean();
+                long[] threadIds = threadBean.findDeadlockedThreads();
+                if (threadIds != null && threadIds.length > 0) {
+                    ThreadInfo[] threadInfos = threadBean.getThreadInfo(threadIds, false,
+                            false);
+                    String threadNames = Arrays.stream(threadInfos)
+                            .map(threadInfo -> threadInfo.getThreadName()
+                                    + "(tid=" + threadInfo.getThreadId() + ")")
+                            .collect(Collectors.joining(", "));
+                    if (System.currentTimeMillis() - threadDumpLoggedTimestamp
+                            > LOG_THREADDUMP_INTERVAL_WHEN_DEADLOCK_DETECTED) {
+                        String diagnosticResult = ThreadDumpUtil.buildThreadDiagnosticString();
+                        log.error("Deadlock detected, service may be unavailable, "
+                                + "thread stack details are as follows: {}.", diagnosticResult);
+                        threadDumpLoggedTimestamp = System.currentTimeMillis();
+                    } else {
+                        log.error("Deadlocked threads detected. {}", threadNames);
+                    }
```

Comment:
Since the status endpoint doesn't have authentication, it will be necessary to have a solution to prevent introducing a new DoS vulnerability where calling the status endpoint in a tight loop could introduce significant load to the system. One way would be to check that the deadlock check is executed only when there's more than 1 seconds from the previous check. If it's less than that, the previous result of the deadlock check would be reused.

Reply:
> Since the status endpoint doesn't have authentication, it will be necessary to have a solution to prevent introducing a new DoS vulnerability where calling the status endpoint in a tight loop could introduce significant load to the system. One way would be to check that the deadlock check is executed only when there's more than 1 seconds from the previous check. If it's less than that, the previous result of the deadlock check would be reused.

This is a very good proposal, and there is indeed a risk of being attacked by DoS. The code logic has been adjusted accordingly. Please help review the code again.